From 0f6409e727b48c7ddc544f61c437a84f942cdee2 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Mon, 13 May 2019 12:56:36 +0200
Subject: [PATCH 1/1] SctPkg: Init: use EFI_UNICODE_COLLATION_PROTOCOL2_GUID

EFI_UNICODE_COLLATION_PROTOCOL_GUID is not defined in the current UEFI
specification. Use  EFI_UNICODE_COLLATION_PROTOCOL2_GUID instead when
initializing the SCT.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 .../SctPkg/Library/SctGuidLib/SctGuidLib.c    |  1 +
 uefi-sct/SctPkg/Library/SctLib/Guid.c         |  1 +
 uefi-sct/SctPkg/Library/SctLib/Init.c         | 21 ++++++++-----------
 uefi-sct/SctPkg/Library/SctLib/SctLib.inf     |  1 +
 4 files changed, 12 insertions(+), 12 deletions(-)

diff --git a/uefi-sct/SctPkg/Library/SctGuidLib/SctGuidLib.c b/uefi-sct/SctPkg/Library/SctGuidLib/SctGuidLib.c
index cb5a7c97..7b230a49 100644
--- a/uefi-sct/SctPkg/Library/SctGuidLib/SctGuidLib.c
+++ b/uefi-sct/SctPkg/Library/SctGuidLib/SctGuidLib.c
@@ -39,6 +39,7 @@ EFI_GUID gEfiDeviceIoProtocolGuid = { 0xAF6AC311, 0x84C3, 0x11D2, { 0x8E, 0x3C,
 EFI_GUID gEfiFileInfoGuid = { 0x09576E92, 0x6D3F, 0x11D2, { 0x8E, 0x39, 0x00, 0xA0, 0xC9, 0x69, 0x72, 0x3B }};
 EFI_GUID gEfiFileSystemInfoGuid = { 0x09576E93, 0x6D3F, 0x11D2, { 0x8E, 0x39, 0x00, 0xA0, 0xC9, 0x69, 0x72, 0x3B }};
 EFI_GUID gEfiUnicodeCollationProtocolGuid = { 0x1D85CD7F, 0xF43D, 0x11D2, { 0x9A, 0x0C, 0x00, 0x90, 0x27, 0x3F, 0xC1, 0x4D }};
+EFI_GUID gEfiUnicodeCollation2ProtocolGuid = { 0xA4C751FC, 0x23AE, 0x4C3E, { 0x92, 0xE9, 0x49, 0x64, 0xCF, 0x63, 0xF3, 0x49 }};
 EFI_GUID gEfiSerialIoProtocolGuid = { 0xBB25CF6F, 0xF1D4, 0x11D2, { 0x9A, 0x0C, 0x00, 0x90, 0x27, 0x3F, 0xC1, 0xFD }};
 EFI_GUID gEfiSimpleNetworkProtocolGuid = { 0xA19832B9, 0xAC25, 0x11D3, { 0x9A, 0x2D, 0x00, 0x90, 0x27, 0x3F, 0xC1, 0x4D }};
 EFI_GUID gEfiNetworkInterfaceIdentifierProtocolGuid = { 0xE18541CD, 0xF755, 0x4F73, { 0x92, 0x8D, 0x64, 0x3C, 0x8A, 0x79, 0xB2, 0x29 }};
diff --git a/uefi-sct/SctPkg/Library/SctLib/Guid.c b/uefi-sct/SctPkg/Library/SctLib/Guid.c
index d0391060..26fcf738 100644
--- a/uefi-sct/SctPkg/Library/SctLib/Guid.c
+++ b/uefi-sct/SctPkg/Library/SctLib/Guid.c
@@ -118,6 +118,7 @@ struct {
   { &gEfiFileSystemInfoGuid,            L"FileSysInfo" },
 
   { &gEfiUnicodeCollationProtocolGuid,  L"UnicodeCollation" },
+  { &gEfiUnicodeCollation2ProtocolGuid, L"UnicodeCollation2" },
   { &gEfiSerialIoProtocolGuid,          L"serialio" },
   { &gEfiSimpleNetworkProtocolGuid,     L"net" },
   { &gEfiNetworkInterfaceIdentifierProtocolGuid,    L"nii" },
diff --git a/uefi-sct/SctPkg/Library/SctLib/Init.c b/uefi-sct/SctPkg/Library/SctLib/Init.c
index 7e7618eb..a8537018 100644
--- a/uefi-sct/SctPkg/Library/SctLib/Init.c
+++ b/uefi-sct/SctPkg/Library/SctLib/Init.c
@@ -15,11 +15,6 @@
 
 #include "SctLibInternal.h"
 
-//
-// International Language
-//
-#define ISO_639_2_ENTRY_SIZE    3
-
 #include EFI_PROTOCOL_DEFINITION (LoadedImage)
 
 //
@@ -91,7 +86,7 @@ InitializeUnicodeSupport (
   //
 
   for (Index=0; Index < NoHandles; Index++) {
-    Status = tBS->HandleProtocol (Handles[Index], &gEfiUnicodeCollationProtocolGuid, (VOID**)&Ui);
+    Status = tBS->HandleProtocol (Handles[Index], &gEfiUnicodeCollation2ProtocolGuid, (VOID**)&Ui);
     if (EFI_ERROR(Status)) {
       continue;
     }
@@ -101,16 +96,18 @@ InitializeUnicodeSupport (
     //
 
     Languages = Ui->SupportedLanguages;
-    Length = SctAsciiStrLen (Languages);
-    for (Position=0; Position < Length; Position += ISO_639_2_ENTRY_SIZE) {
+    Length = SctAsciiStrLen (Languages) - SctAsciiStrLen (LangCode);
+    for (Position=0; Position <= Length; ++Position) {
 
       //
       // If this code matches, use this driver
       //
 
-      if (SctCompareMem (Languages+Position, LangCode, ISO_639_2_ENTRY_SIZE) == 0) {
-        UnicodeInterface = Ui;
-        goto Done;
+      if (!Position || Languages[Position - 1] == ';') {
+        if (SctCompareMem (Languages+Position, LangCode, SctAsciiStrLen (LangCode)) == 0) {
+          UnicodeInterface = Ui;
+          goto Done;
+        }
       }
     }
   }
@@ -180,7 +177,7 @@ SctInitializeLib (
     //
     // LangCode = LibGetVariable (VarLanguage, &EfiGlobalVariable);
     // InitializeUnicodeSupport (LangCode);
-    InitializeUnicodeSupport ("eng");
+    InitializeUnicodeSupport ("en");
     if (LangCode) {
       SctFreePool (LangCode);
     }
diff --git a/uefi-sct/SctPkg/Library/SctLib/SctLib.inf b/uefi-sct/SctPkg/Library/SctLib/SctLib.inf
index 6db9ae09..5af8728e 100644
--- a/uefi-sct/SctPkg/Library/SctLib/SctLib.inf
+++ b/uefi-sct/SctPkg/Library/SctLib/SctLib.inf
@@ -116,6 +116,7 @@
   gEfiSimpleTextInProtocolGuid
   gEfiSimpleTextOutProtocolGuid
   gEfiUnicodeCollationProtocolGuid
+  gEfiUnicodeCollation2ProtocolGuid
   gEfiDriverBindingProtocolGuid
   gEfiDriverConfigurationProtocolGuid
   gEfiDriverDiagnosticsProtocolGuid
-- 
2.20.1

